website/pages/blog/building-cloudquery.md

```diff
@@ -0,0 +1,104 @@
+
+### Code Generation
+
+A big challenge is scaling and continuously supporting a large number of APIs. Before we dive into how we leveraged code-generation to generate CloudQuery plugins, let’s take a quick look at how client library generation for multiple languages often work.
+
+Usually there is some intermediate language such as gRPC/GraphQL/Smithy that generates server stubs and clients automatically for multiple languages instead of manually maintaining clients for a number of languages and keeping up with the server APIs.
+
+There were some significant advances around client library generation in the recent years that we also took advantage of.
+
+We could take advantage of the schema definition and generate the source plugin from that, but we decided to do that one step further in the pipeline: from the Go Code.
+
+Go is incredibly suited for code analysis so we created a [code generation library](https://github.com/cloudquery/plugin-sdk/tree/main/codegen) that generates CloudQuery schema from Go structs using reflection. This helps achieve number of things:
+
+1. Keep CloudQuery schema up-to-date with APIs automatically
+2. Enable developers to use the Go Client which is built for developers and include important capabilities when talking to the service API.
+3. Our current set of plugins are >80% auto-generated!
+
+## Scaling Destinations
+
+### Pluggable Destinations
+
+Another key requirement is supporting multiple destinations. The number of destinations is actually limited, but if destinations won't be decoupled from sources, source plugins will need to be updated every time there is a new destination, which will require a growing `n*m` (`n` number of source plugins and `m` number of destinations plugins) number of code-changes and make development unsustainable.
+
+Destination plugins are implemented in a similar way to source plugins as gRPC servers.
+
+### Typing System
+
+The job of an ELT platform is to mostly do two things:
+
+1. Transform and normalize the data from an API
+2. For each transformed field add type data.
+
+CloudQuery transforms every single field to its own rich type system, which contains more than [17 types](https://github.com/cloudquery/plugin-sdk/tree/main/schema) (including things like IP Addresses, MAC Addresses) This ensures all the validation is happening in the transformation phase and then all destination plugins needs to do is just transform it depending on what types each destination is supporting. This is a big shift from how Singer or Airbyte works where they mostly use what is available in JSON and [JSON Schema](https://json-schema.org/) as they couple the encoding together with their type system.
+
+With those two decisions CloudQuery already supports more than [5 destinations](https://www.cloudquery.io/docs/plugins/destinations) two months after our initial [V1 release](https://www.cloudquery.io/blog/cloudquery-v1-release)
```

Comment:
```suggestion
Because of these two design decisions, CloudQuery already supports more than [3 destinations](https://www.cloudquery.io/docs/plugins/destinations) only two months after our initial [V1 release](https://www.cloudquery.io/blog/cloudquery-v1-release)
```

Unless we're planning on releasing some more before this blog post is released? 😄

Reply:
There are 5 actually, PostgreSQL, Snowflake, CockroachDB, CSV, SQLite

Reply:
Ah yes, I forgot about CockroachDB (maybe it would be nice to list it here under an asterisk for postrgesql: https://www.cloudquery.io/docs/plugins/destinations/overview)

In any case, I've updated my suggestion to say "supports 5 destinations" instead of "supports more than 5 destinations"